File: src/oca/java/test/UserTest.java
/*******************************************************************************
 * Copyright 2002-2016, OpenNebula Project, OpenNebula Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ******************************************************************************/
import static org.junit.Assert.*;

import javax.xml.xpath.XPath;
import javax.xml.xpath.XPathExpressionException;
import javax.xml.xpath.XPathFactory;

import org.junit.After;
import org.junit.AfterClass;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Test;
import org.opennebula.client.Client;
import org.opennebula.client.OneResponse;
import org.opennebula.client.OneSystem;
import org.opennebula.client.user.User;
import org.opennebula.client.user.UserPool;
import org.opennebula.client.group.Group;
import org.w3c.dom.Node;

public class UserTest
{

    private static User     user;
    private static UserPool userPool;

    private static Client client;

    private static OneResponse  res;
    private static String       name      = "new_test_user";
    private static String       password  = "new_test_password";

    /**
     * @throws java.lang.Exception
     */
    @BeforeClass
    public static void setUpBeforeClass() throws Exception
    {
        client      = new Client();
        userPool    = new UserPool(client);
    }

    /**
     * @throws java.lang.Exception
     */
    @AfterClass
    public static void tearDownAfterClass() throws Exception
    {
    }

    /**
     * @throws java.lang.Exception
     */
    @Before
    public void setUp() throws Exception
    {
        res = User.allocate(client, name, password);

        assertTrue( res.getErrorMessage(), !res.isError() );

        int uid = Integer.parseInt(res.getMessage());
        user    = new User(uid, client);
    }

    /**
     * @throws java.lang.Exception
     */
    @After
    public void tearDown() throws Exception
    {
        user.delete();
    }


    @Test
    public void allocate()
    {
        userPool.info();

        boolean found = false;
        for(User u : userPool)
        {
            found = found || u.getName().equals(name);
        }

        assertTrue( found );
    }

    @Test
    public void allocateInGroup()
    {
        user.delete();

        res = Group.allocate(client, "group_a");
        assertTrue( res.getErrorMessage(), !res.isError() );
        int group_a_id = Integer.parseInt(res.getMessage());
        Group group_a = new Group(group_a_id, client);

        res = Group.allocate(client, "group_b");
        assertTrue( res.getErrorMessage(), !res.isError() );
        int group_b_id = Integer.parseInt(res.getMessage());
        Group group_b = new Group(group_b_id, client);

        res = Group.allocate(client, "group_c");
        assertTrue( res.getErrorMessage(), !res.isError() );
        int group_c_id = Integer.parseInt(res.getMessage());
        Group group_c = new Group(group_c_id, client);

        Integer[] gids = {group_b_id, group_a_id};
        res = User.allocate(client, "test_user_in_group", "pass", "", gids);

        assertTrue( res.getErrorMessage(), !res.isError() );

        user = new User(Integer.parseInt(res.getMessage()), client);

        res = user.info();
        assertTrue( res.getErrorMessage(), !res.isError() );

        assertTrue( user.gid() == group_b_id );

        group_a.info();
        group_b.info();
        group_c.info();

        assertTrue( group_a.contains(user.id()) );
        assertTrue( group_b.contains(user.id()) );
        assertFalse( group_c.contains(user.id()) );
    }

    @Test
    public void info()
    {
        res = user.info();
        assertTrue( res.getErrorMessage(), !res.isError() );

        assertTrue( user.id() >= 0 );
        assertTrue( user.getName().equals(name) );
    }

    @Test
    public void attributes()
    {
        res = user.info();
        assertTrue( res.getErrorMessage(), !res.isError() );

        assertTrue( user.xpath("NAME").equals(name) );
        assertTrue( user.xpath("ENABLED").equals("1") );
    }

    @Test
    public void chauth()
    {
        res = user.info();
        assertTrue( res.getErrorMessage(), !res.isError() );

        assertTrue( user.xpath("AUTH_DRIVER").equals("core") );

        res = user.chauth("new_driver", password);

        res = user.info();
        assertTrue( res.getErrorMessage(), !res.isError() );

        assertTrue( user.xpath("AUTH_DRIVER").equals("new_driver") );
    }

    @Test
    public void update()
    {
        String new_template =  "ATT2 = NEW_VAL\n" +
                "ATT3 = VAL3";

        res = user.update(new_template);
        assertTrue( res.getErrorMessage(), !res.isError() );

        res = user.info();
        assertTrue( res.getErrorMessage(), !res.isError() );
        assertTrue( user.xpath("TEMPLATE/ATT1").equals( "" ) );
        assertTrue( user.xpath("TEMPLATE/ATT2").equals( "NEW_VAL" ) );
        assertTrue( user.xpath("TEMPLATE/ATT3").equals( "VAL3" ) );
    }

    @Test
    public void delete()
    {
        res = user.info();
        assertTrue( res.getErrorMessage(), !res.isError() );
        assertTrue( user.isEnabled() );

        res = user.delete();
        assertTrue( res.getErrorMessage(), !res.isError() );

        res = user.info();
        assertTrue( res.getErrorMessage(), res.isError() );
    }

    @Test
    public void defaultqutoas()
    {
        OneSystem system = new OneSystem(client);

        res = system.getUserQuotas();
        assertTrue( res.getErrorMessage(), !res.isError() );

        res = system.setUserQuotas("VM = [ VMS = 7, MEMORY = 0, CPU = 3, SYSTEM_DISK_SIZE = 1 ]");
        assertTrue( res.getErrorMessage(), !res.isError() );

        Node node = system.getUserQuotasXML();
        XPathFactory factory = XPathFactory.newInstance();
        XPath xpath = factory.newXPath();

        try
        {
            assertTrue( xpath.evaluate("VM_QUOTA/VM/VMS", node).equals("7") );
        } catch (XPathExpressionException e)
        {
            assertTrue(e.getMessage(), false);
        }
    }
}

File: src/oca/go/src/goca/user_test.go
/* -------------------------------------------------------------------------- */
/* Copyright 2002-2025, OpenNebula Project, OpenNebula Systems                */
/*                                                                            */
/* Licensed under the Apache License, Version 2.0 (the "License"); you may    */
/* not use this file except in compliance with the License. You may obtain    */
/* a copy of the License at                                                   */
/*                                                                            */
/* http://www.apache.org/licenses/LICENSE-2.0                                 */
/*                                                                            */
/* Unless required by applicable law or agreed to in writing, software        */
/* distributed under the License is distributed on an "AS IS" BASIS,          */
/* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   */
/* See the License for the specific language governing permissions and        */
/* limitations under the License.                                             */
/*--------------------------------------------------------------------------- */

package goca

import (
	"github.com/OpenNebula/one/src/oca/go/src/goca/parameters"
	. "gopkg.in/check.v1"
)

type UserSuite struct {
	groupID int
	userID  int
}

var _ = Suite(&UserSuite{})

func (s *UserSuite) SetUpSuite(c *C) {
	// Create Group
	id, err := testCtrl.Groups().Create(GenName("TestGroup"))
	c.Assert(err, IsNil)
	s.groupID = id
}

func (s *UserSuite) TearDownSuite(c *C) {
	// Delete Group
	groupC := testCtrl.Group(s.groupID)
	groupC.Delete()
}

func (s *UserSuite) SetUpTest(c *C) {
	// Create User
	id, err := testCtrl.Users().Create("TestUser", "passwd", "", []int{})
	c.Assert(err, IsNil)
	s.userID = id
}

func (s *UserSuite) TearDownTest(c *C) {
	// Delete User
	user := testCtrl.User(s.userID)
	user.Delete()
}

func (s *UserSuite) TestGetByNameAndID(c *C) {
	// Get User by ID
	user, err := testCtrl.User(s.userID).Info(false)

	c.Assert(err, IsNil)
	c.Assert(user.ID, Equals, s.userID)

	// Test value
	c.Assert(user.GName, Equals, "users")

	// Get User by name
	id, err := testCtrl.Users().ByName(user.Name)
	c.Assert(err, IsNil)
	c.Assert(user.ID, Equals, id)
}

func (s *UserSuite) TestUpdate(c *C) {
	userC := testCtrl.User(s.userID)
	err := userC.Update(`DESCRIPTION="Test group description"`, parameters.Merge)

	c.Assert(err, IsNil)

	user, err := testCtrl.User(s.userID).Info(false)
	c.Assert(err, IsNil)

	description, err := user.Template.GetStr("DESCRIPTION")
	c.Assert(err, IsNil)
	c.Assert(description, Equals, "Test group description")
}

func (s *UserSuite) TestLogin(c *C) {
	userC := testCtrl.User(s.userID)
	err := userC.Login("", 30, 1)
	c.Assert(err, IsNil)
}

func (s *UserSuite) TestAddChangeDelGroup(c *C) {
	// Add Group
	userC := testCtrl.User(s.userID)
	err := userC.AddGroup(s.groupID)
	c.Assert(err, IsNil)

	// Change Group
	err = userC.Chgrp(s.groupID)
	c.Assert(err, IsNil)

	// Remove Group
	err = userC.DelGroup(1)
	c.Assert(err, IsNil)
}

func (s *UserSuite) TestChauth(c *C) {
	userC := testCtrl.User(s.userID)
	err := userC.Chauth("dummy", "dummy_passwd")
	c.Assert(err, IsNil)
}

func (s *UserSuite) TestEnable(c *C) {
	// Disable user
	userC := testCtrl.User(s.userID)
	err := userC.Enable(false)
	c.Assert(err, IsNil)

	// Check user disabled
	user, err := testCtrl.User(s.userID).Info(false)
	c.Assert(err, IsNil)
	c.Assert(user.Enabled, Equals, 0)

	// Enable user
	err = userC.Enable(true)
	c.Assert(err, IsNil)
}

func (s *UserSuite) TestPassword(c *C) {
	userC := testCtrl.User(s.userID)
	err := userC.Passwd("new_password")
	c.Assert(err, IsNil)
}

func (s *UserSuite) TestDefaultQuota(c *C) {
	// Set default quota
	quotaTpl := `VM = [ VMS = "5" ]`
	err := testCtrl.Users().DefaultQuotaUpdate(quotaTpl)
	c.Assert(err, IsNil)

	// Check if the quota is set correctly
	quota, err := testCtrl.Users().DefaultQuotaInfo()
	c.Assert(err, IsNil)

	vms := quota.VM[0].VMs
	c.Assert(vms, Equals, 5)

	// Delete default quota
	err = testCtrl.Users().DefaultQuotaUpdate(`VM = [ VMS = "-2" ]`)
	c.Assert(err, IsNil)
}

func (s *UserSuite) TestQuota(c *C) {
	quotaTpl := `VM = [ VMS = "10" ]`
	userC := testCtrl.User(s.userID)
	err := userC.Quota(quotaTpl)
	c.Assert(err, IsNil)

	// Check quota added
	group, err := testCtrl.User(s.userID).Info(false)
	c.Assert(err, IsNil)

	quota := group.QuotasList.VM[0].VMs
	c.Assert(quota, Equals, 10)
}
